### DISCUSSION TITLE
How to compose batch object

### DISCUSSION THREAD

**Discussion by lukyer** [open-policy-agent/community#244] [OPA and Rego] [answered]
Hi all,
I'm fighting Rego to achieve simple bulk operation (i.e. evaluate permissions for these 1000 entities at once). How can I compose this resulting object *regardless some rule evaluation is FALSE* ? The falsy rule always terminate the main flow. Basically I need standard `else` branching in rules 😄  

I have also tried comprehension with no luck. Another thing I tried is to compose several objects with truthy logic only and try to merge it in the end but merge/map is not easily doable using std lib.

I must have been missing something basic 😄 

Minimal example: https://play.openpolicyagent.org/p/uPp6OmMyx0

Thanks!

**Comment by anderseknert** [comment, marked as answer]
Hi @lukyer! 

If no return value is defined for a rule or function, Rego will consider the result to be _undefined_ on failure (i.e. if not all of the conditions in the body are true). As you noted, encountering undefined stops policy evaluation. When working with boolean rules (e.g. "`allow`", etc) it's a comon practice to assign a `default` value of `false` to the rule so that failure still evalutes to _something_, whiich would be boolean `false`.

Functions cannot [yet](https://github.com/open-policy-agent/opa/issues/2445) have a default value assigned, so you'll commonly see either another function declaration to handle the "other" case, or simply an `else` clause to deal with the failure case. 

Your example using `else` to ensure the functions return `false` rather than undefined: https://play.openpolicyagent.org/p/P7U3SyohSq

**Comment by lukyer** [comment]
@anderseknert thanks! That is the `else` syntax for functions I was missing 😄  

I have tried the `default` value for functions as you noted with no luck but I haven't discovered this trick. It would be great to document it [here](https://www.openpolicyagent.org/docs/latest/policy-language/#functions) and/or [here](https://www.openpolicyagent.org/docs/latest/policy-language/#else-keyword).

**Reply by anderseknert** [reply]
Yeah, at least a pointer to the docs on `else` from that first link (the section on functions) would be good, I think. If you'd want to create an issue — or even a PR — to have that added, that'd be great 🙂

**Reply by lukyer** [reply]
https://github.com/open-policy-agent/opa/pull/4996

**Reply by anderseknert** [reply]
Nice, thanks 👍
